flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,315 @@
+    */
+  override def total: Int = data.iterator.map(_._2).sum
+
+  /** Bin counter access function
+    *
+    * @param bin bin number to access
+    * @return `m_bin` = counter of bin
+    */
+  private[statistics] def getCounter(bin: Int): Int = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._2
+  }
+
+  /** Number of bins currently being used
+    *
+    * @return number of bins
+    */
+  override def bins: Int = {
+    data.size
+  }
+
+  /** Returns the string representation of the histogram.
+    *
+    */
+  override def toString: String = {
+    s"Size:" + bins + " " + data.toString
+  }
+
+  /** Loads values and counters into the histogram.
+    * This action can only be performed when there the histogram is empty
+    *
+    * @param counts Array of tuple of (value,count)
+    */
+  def loadData(counts: Array[(Double, Int)]): Unit = {
+    if (data.nonEmpty) {
+      throw new RuntimeException("Data can only be loaded during initialization")
+    }
+    data ++= counts
+    require(checkSanity, "Invalid initialization from data")
+  }
+
+  /** Bin value access function
+    *
+    * @param bin bin number to access
+    * @return `v_bin` = value of bin
+    */
+  private[statistics] def getValue(bin: Int): Double = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._1
+  }
+
+  /** Checks whether the values loaded into data by loadData are properly sorted or not.
+    * All values should be in (lower,upper)
+    * All counters loaded into data by loadData should be positive
+    *
+    */
+  private def checkSanity: Boolean = {
+    val size = bins
+    if (size == 0) return true
+    if (lower >= getValue(0)) return false
+    if (upper <= getValue(size - 1)) return false
+    for (i <- 0 to size - 2) {
+      if (getValue(i + 1) < getValue(i)) return false // equality will get merged later on
+      if (getCounter(i) <= 0) return false
+    }
+    if (getCounter(size - 1) <= 0) return false
+    // now bring the histogram to capacity and merge all <i>very-close</i> elements
+    while (mergeElements()) {}
+    true
+  }
+
+  /** Searches for the index with value just greater than p
+    * If `p >= v_{bins-1}`, return bins
+    *
+    * @param p value to search for
+    * @return the bin with value just greater than p
+    */
+  private def find(p: Double): Int = {
+    val size: Int = bins
+    if (size == 0 || p < getValue(0)) {
+      return 0
+    }
+    for (i <- 0 to size - 2) {
+      if (p >= getValue(i) && p < getValue(i + 1)) {
+        return i + 1
+      }
+    }
+    size
+  }
+
+  /** Merges the closest two elements in the histogram
+    * If we're over capacity, merge without thought
+    * Otherwise, only merge when two elements have almost equal values. We're talking doubles, so
+    * no exact equalities
+    *
+    * @return Whether we succeeded in merging any two elements
+    */
+  private def mergeElements(): Boolean = {
```

Comment:
Calling this method only merges two bins together. If you have to merge multiple bins because you're exceeding the capacity by more than 1, then you have to recompute the difference values over and over again even though they don't really change (only for two bin pairs). This is highly inefficient. Can't we write a method which merges elements until the merge condition is false? While merging, this method can reuse the calculated difference values.